### ISSUE TITLE
[Feature Request]: Support object tree shaking 

### ISSUE THREAD

**Issue by lishaduck** [rolldown/rolldown#2995]
### What problem does this feature solve?

I switched from using `require()` to load a `package.json` to a static `import`. This inlined the JSON (fine by me; I would've prefered externalizing it but putting `./package.json` as external didn't seem to work), which enlarged the file significantly. I don't want to it bundle all of the file, just what's used (the `version` field). Using a namespace import worked, but Node.js stabilized JSON modules without named exports, just a default export, so I'd prefer to use that.

### What does the proposed API look like?

Rollup recently landed object tree shaking (though it was reverted), which seems like a more correct solution than special-casing JSON default import that are only used statically.

**Comment by hyf0** [comment]
This should be supported. Could you create a minimal reproduction for your case in https://stackblitz.com/github/rolldown/rolldown-starter-stackblitz?  @IWANABETHATGUY cc

**Comment by IWANABETHATGUY** [comment]
I think this is related to https://github.com/rolldown/rolldown/pull/3003, seems the wrapkind detector is wrong, we think `test.json` `module.exportsKind = Esm` even `test.json` is `require` by entry.

**Comment by IWANABETHATGUY** [comment]
for esbuild, it is correct determined, https://hyrious.me/esbuild-repl/?version=0.23.0&b=e%00entry.js%00const+a+%3D+require%28%27.%2Fpackage.json%27%29%0Aconsole.log%28a.a%29&b=%00test.json%00&b=%00package.json%00%7B%0A++%22type%22%3A+%22module%22%0A%7D&o=%7B%0A++treeShaking%3A+true%2C%0A+%0A%22bundle%22%3A+true%2C%0Aformat%3A+%22esm%22%2C%0Aminify%3A+false%2C%0Adefine%3A+%7B%0A++%22a.b%22%3A+%22a.c%22%0A%7D%0A%7D.

**Comment by IWANABETHATGUY** [comment]
If we remove `type: module` in `rolldown/examples/basic/package.json` the output should same as esbuild.

**Comment by IWANABETHATGUY** [comment]
## Output when have `type: module` in package.json
<summary>
 <details>

```js

//#region rolldown:runtime
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __esm = (fn, res) => function() {
	return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
	for (var name in all) __defProp(target, name, {
		get: all[name],
		enumerable: true
	});
};
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

//#endregion
//#region package.json
var package_exports = {};
__export(package_exports, {
	default: () => package_default,
	type: () => type
});
var type, package_default;
var init_package = __esm({ "package.json"() {
	type = "module";
	package_default = { type };
} });

//#endregion
//#region entry.js
const a = (init_package(), __toCommonJS(package_exports).default);
console.log(`a: `, a);

//#endregion

```
</details>
</summary>

## Output without `type: module`
<summary>
<details>

```js

//#region rolldown:runtime
var __getOwnPropNames = Object.getOwnPropertyNames;
var __commonJS = (cb, mod) => function() {
	return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};

//#endregion
//#region package.json
var require_package = __commonJS({ "package.json"(exports, module) {
	module.exports = { "author": {} };
} });

//#endregion
//#region entry.js
const a = require_package();
console.log(`a: `, a);

//#endregion

```
</details>
</summary>

**Comment by IWANABETHATGUY** [comment]
If you require a json module, the whole module will be bundled for now despite the usage, which is the same as esbuild, it is expected for now, since we don't support cjs tree shaking.

**Comment by hyf0** [comment]
> If we remove `type: module` in `rolldown/examples/basic/package.json` the output should same as esbuild.

I guess we should only care about package.json if the module is ecma module. Tricky part is where we break the chain.

**Comment by IWANABETHATGUY** [comment]
Looks this is different compare to esbuild, https://github.com/evanw/esbuild/blob/d34e79e2a998c21bb71d57b92b0017ca11756912/internal/bundler/bundler.go#L1446-L1460, 
ours: 
https://github.com/rolldown/rolldown/blob/e5167f21b7584061fad1ed69be316ee4020765a3/crates/rolldown_resolver/src/resolver.rs?plain=1#L246-L262

**Comment by lishaduck** [comment]
> Could you create a minimal reproduction for your case?

Do you still want one?

**Comment by hyf0** [comment]
> > Could you create a minimal reproduction for your case?
> 
> Do you still want one?

No. We have locate the problem. We will soon raise a fix for this issue.

**Comment by IWANABETHATGUY** [comment]
> > Could you create a minimal reproduction for your case?
> 
> Do you still want one?

I fixed module format infer, but after I review your description, I found that this might not fully same issues as you described, a reproduction demo would be more helpful.

**Comment by lishaduck** [comment]
Ok, made a repro repo: [`github:lishaduck/rolldown-repro`](https://github.com/lishaduck/rolldown-repro).

**Comment by IWANABETHATGUY** [comment]
> Ok, made a repro repo: [`github:lishaduck/rolldown-repro`](https://github.com/lishaduck/rolldown-repro).

This is out of our scope for current stage, maybe we would consider adding it in the future.

**Comment by IWANABETHATGUY** [comment]
As a workaround use `import * as ns` instead, for object tree shaking, you need to store the each object shape, it is complicated and error-prone, so we don't have the plan to support it now.

**Comment by lishaduck** [comment]
I agree that object tree shaking is out of scope for now, but in the meantime, could .json be special-cased so that `import default` is the same as `import * as default`?

**Comment by IWANABETHATGUY** [comment]
it may introduce a lot of hard-coded logic, maybe not be preferable, we will have a look what other bundler's behavior, and decide how to deal with it.

**Comment by sapphi-red** [comment]
linking https://github.com/oxc-project/oxc/issues/10844, https://github.com/oxc-project/oxc/issues/10845

**Comment by lishaduck** [comment]
This was fixed by https://github.com/rolldown/rolldown/pull/6626 for my use case at least.

**Comment by hyfdev** [comment]
Status Update: This feature is planned but require changes on the core architecture. We don't have bandwidth to touch the architecture so far
